os/drivers/lwnl/lwnl.c

```diff
@@ -169,69 +161,55 @@ static int lwnl_close(struct file *filep)
 			ln_open->io_fds[waiter_idx] = NULL;
 		}
 	}
-	LWNLDEV_UNLOCK(upper);
 #endif
+	LWNLDEV_UNLOCK(upper);
 
 	int res = lwnl_remove_listener(filep);
 	if (res < 0) {
 		ret = -ENOSYS;
-		goto errout;
 	}
 
-errout:
 	LWNL_LEAVE;
 	return ret;
 }
 
-
 static ssize_t lwnl_read(struct file *filep, char *buffer, size_t len)
 {
 	LWNL_ENTER;
+
 	int res = lwnl_get_event(filep, buffer, len);
+
 	// todo_net : convert res to vfs error style?
 	LWNL_LEAVE;
 	return res;
 }
 
-#ifndef CONFIG_NET_NETMGR
-extern int lwnl_message_handle(const char *msg, int msg_len);
-extern void lwnl_initialize_dev(void);
-#else
-extern int netdev_req_handle(const char *msg, size_t msg_len);
-#endif
-
 static ssize_t lwnl_write(struct file *filep, const char *buffer, size_t len)
 {
 	LWNL_ENTER;
-	int ret = -EINVAL;
-	struct inode *inode = filep->f_inode;
-	struct lwnl_upperhalf_s *upper = inode->i_private;
-
-	LWNLDEV_LOCK(upper);
-
 #ifdef CONFIG_NET_NETMGR
-	ret = netdev_req_handle(buffer, len);
+	int ret = netdev_req_handle(buffer, len);
 #else
-	ret = lwnl_message_handle(buffer, len);
+	int ret = lwnl_message_handle(buffer, len);
 #endif
-
-	LWNLDEV_UNLOCK(upper);
 	LWNL_LEAVE;
 	if (ret < 0) {
 		return -1;
 	}
 	return len;
 }
 
-
 static int lwnl_ioctl(struct file *filep, int cmd, unsigned long arg)
 {
 	LWNL_ENTER;
-
-	return 0;
+	int res = lwnl_add_listener(filep);
+	if (res < 0) {
+		res = -ENOSYS;
```

Comment:
same as above. if there's no implemented code in ioctl , then it should return -ENOSYS
if lwnl_add_listener has no implemented code, it should return -ENOSYS
so res will be smaller than 0, it doesn't affect here.